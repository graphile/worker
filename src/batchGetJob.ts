import {
  CompiledSharedOptions,
  EnhancedWithPgClient,
  WorkerPoolOptions,
} from ".";
import { GetJobFunction, Job, TaskList, WorkerPool } from "./interfaces";
import { getJob as baseGetJob } from "./sql/getJob";

export function makeBatchGetJob(
  compiledSharedOptions: CompiledSharedOptions<WorkerPoolOptions>,
  tasks: TaskList,
  withPgClient: EnhancedWithPgClient,
  workerPool: WorkerPool,
  getJobBatchSize: number,
): GetJobFunction {
  let getJobCounter = 0;
  let jobQueue: Job[] = [];
  let nextJobs: Promise<boolean> | null = null;
  let getJobBaseline = 0;
  const getJob: GetJobFunction = async (workerId, flagsToSkip) => {
    // Cannot batch if there's flags
    if (flagsToSkip !== null) {
      const jobs = await baseGetJob(
        compiledSharedOptions,
        withPgClient,
        tasks,
        workerPool.id,
        flagsToSkip,
        1,
      );
      return jobs[0];
    }

    const job = jobQueue.pop();
    if (job !== undefined) {
      return job;
    } else {
      return batchGetJob(++getJobCounter);
    }
  };

  const batchGetJob = async (myFetchId: number): Promise<Job | undefined> => {
    // TODO rewrite this so that if we have batch size of 1 we'll still fetch newer jobs in parallel (not queued)
    if (!nextJobs) {
      // Queue is empty, no fetch of jobs in progress; let's fetch them.
      getJobBaseline = getJobCounter;
      nextJobs = (async () => {
        try {
          const jobs = await baseGetJob(
            compiledSharedOptions,
            withPgClient,
            tasks,
            workerPool.id,
            null,
            getJobBatchSize,
          );
          jobQueue = jobs.reverse();
          return jobs.length >= getJobBatchSize;
        } finally {
          nextJobs = null;
        }
      })();
    }
    const fetchedMax = await nextJobs;
    const job = jobQueue.pop();
    if (job) {
      return job;
    } else if (fetchedMax || myFetchId > getJobBaseline) {
      // Either we fetched as many jobs as we could and there still weren't
      // enough, or we requested a job after the request for jobs was sent to
      // the database. Either way, let's fetch again.
      return batchGetJob(myFetchId);
    } else {
      return undefined;
    }
  };

  return getJob;
}
