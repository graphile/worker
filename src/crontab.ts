import * as JSON5 from "json5";
import { parse } from "querystring";

import {
  CRONTAB_COMMAND,
  CRONTAB_LINE_PARTS,
  CRONTAB_OPTIONS_BACKFILL,
  CRONTAB_OPTIONS_ID,
  CRONTAB_OPTIONS_MAX,
  CRONTAB_OPTIONS_PRIORITY,
  CRONTAB_OPTIONS_QUEUE,
  CRONTAB_TIME_PARTS,
  PERIOD_DURATIONS,
  TIMEPHRASE_PART,
} from "./cronConstants";
import { parseCrontabRanges } from "./cronMatcher";
import { CronItem, CronItemOptions, ParsedCronItem } from "./interfaces";

/**
 * Returns a period of time in milliseconds representing the time phrase given.
 *
 * Time phrases are comprised of a sequence of number-letter combinations,
 * where the number represents a quantity and the letter represents a time
 * period, e.g.  `5d` for `five days`, or `3h` for `three hours`; e.g.
 * `4w3d2h1m` represents `four weeks, three days, 2 hours and 1 minute` (i.e. a
 * period of 44761 minutes).  The following time periods are supported:
 *
 * - `s` - one second (1000 milliseconds)
 * - `m` - one minute (60 seconds)
 * - `h` - one hour (60 minutes)
 * - `d` - on day (24 hours)
 * - `w` - one week (7 days)
 */
const parseTimePhrase = (timePhrase: string): number => {
  let remaining = timePhrase;
  let milliseconds = 0;
  while (remaining.length) {
    const matches = TIMEPHRASE_PART.exec(remaining);
    if (!matches) {
      throw new Error(
        `Invalid time phrase '${timePhrase}', did not understand '${remaining}'`,
      );
    }
    const [wholeMatch, quantity, period] = matches;
    const periodDuration = PERIOD_DURATIONS[period] || 0;
    milliseconds += parseInt(quantity, 10) * periodDuration;
    remaining = remaining.slice(wholeMatch.length);
  }
  return milliseconds;
};

const parseCrontabOptions = (
  lineNumber: number,
  optionsString: string | undefined,
): { options: CronItemOptions; identifier: string | undefined } => {
  const parsed = optionsString != null ? parse(optionsString) : {};
  let backfillPeriod: number | undefined = undefined;
  let maxAttempts: number | undefined = undefined;
  let identifier: string | undefined = undefined;
  let queueName: string | undefined = undefined;
  let priority: number | undefined = undefined;

  type MatcherTuple = [RegExp, (matches: RegExpExecArray) => void];

  const matchers: { [key: string]: MatcherTuple } = {
    id: [
      CRONTAB_OPTIONS_ID,
      (matches) => {
        identifier = matches[1];
      },
    ],
    fill: [
      CRONTAB_OPTIONS_BACKFILL,
      (matches) => {
        backfillPeriod = parseTimePhrase(matches[1]);
      },
    ],
    max: [
      CRONTAB_OPTIONS_MAX,
      (matches) => {
        maxAttempts = parseInt(matches[1], 10);
      },
    ],
    queue: [
      CRONTAB_OPTIONS_QUEUE,
      (matches) => {
        queueName = matches[1];
      },
    ],
    priority: [
      CRONTAB_OPTIONS_PRIORITY,
      (matches) => {
        priority = parseInt(matches[1], 10);
      },
    ],
  };

  function match(matcher: MatcherTuple, key: string, value: string) {
    const [regex, set] = matcher;
    const matches = regex.exec(value);
    if (matches) {
      set(matches);
    } else {
      throw new Error(
        `Options on line ${lineNumber} of crontab contains invalid value for '${key}', value '${value}' is not compatible with this option.`,
      );
    }
  }

  Object.entries(parsed).forEach(([key, value]) => {
    if (typeof value !== "string") {
      throw new Error(
        `Options on line ${lineNumber} of crontab contains invalid value for '${key}', did you specify it more than once?`,
      );
    }
    const matcher = Object.prototype.hasOwnProperty.call(matchers, key)
      ? matchers[key]
      : null;
    if (matcher) {
      match(matcher, key, value);
    } else {
      throw new Error(
        `Options on line ${lineNumber} of crontab contains unsupported key '${key}'; supported keys are: '${Object.keys(
          matchers,
        ).join("', '")}'.`,
      );
    }
  });

  if (!backfillPeriod) {
    backfillPeriod = 0;
  }

  return {
    options: { backfillPeriod, maxAttempts, queueName, priority },
    identifier,
  };
};

const parseCrontabPayload = (
  lineNumber: number,
  payloadString: string | undefined,
): any => {
  if (!payloadString) {
    return null;
  }
  try {
    return JSON5.parse(payloadString);
  } catch (e) {
    throw new Error(
      `Failed to parse JSON5 payload on line ${lineNumber} of crontab: ${e.message}`,
    );
  }
};

const parseCrontabCommand = (
  lineNumber: number,
  command: string,
): Pick<ParsedCronItem, "task" | "options" | "payload" | "identifier"> => {
  const matches = CRONTAB_COMMAND.exec(command);
  if (!matches) {
    throw new Error(
      `Invalid command specification in line ${lineNumber} of crontab.`,
    );
  }
  const [, task, optionsString, payloadString] = matches;
  const { options, identifier = task } = parseCrontabOptions(
    lineNumber,
    optionsString,
  );
  const payload = parseCrontabPayload(lineNumber, payloadString);
  return { task, options, payload, identifier };
};

/**
 * Parses a line from a crontab file, such as `* * * * * my_task`
 */
export const parseCrontabLine = (
  crontabLine: string,
  lineNumber: number,
): ParsedCronItem => {
  const matches = CRONTAB_LINE_PARTS.exec(crontabLine);
  if (!matches) {
    throw new Error(
      `Could not process line '${lineNumber}' of crontab: '${crontabLine}'`,
    );
  }
  const { minutes, hours, dates, months, dows } = parseCrontabRanges(
    matches,
    `line ${lineNumber} of crontab`,
  );
  const { task, options, payload, identifier } = parseCrontabCommand(
    lineNumber,
    matches[6],
  );

  return {
    minutes,
    hours,
    dates,
    months,
    dows,
    task,
    options,
    payload,
    identifier,
  };
};

export const parseCrontab = (crontab: string): Array<ParsedCronItem> => {
  const lines = crontab.split(/\r?\n/);
  const items: ParsedCronItem[] = [];
  for (
    let lineNumber = 1, numberOfLines = lines.length;
    lineNumber <= numberOfLines;
    lineNumber++
  ) {
    const line = lines[lineNumber - 1].trim();
    if (line.startsWith("#") || line === "") {
      // Ignore comment lines and empty lines
      continue;
    }
    items.push(parseCrontabLine(line, lineNumber));
  }

  // Assert that identifiers are unique
  const identifiers = items.map((i) => i.identifier);
  identifiers.sort();
  const duplicates = identifiers.filter(
    (id, i) => i > 0 && id === identifiers[i - 1],
  );
  if (duplicates.length) {
    throw new Error(
      `Invalid crontab; duplicate identifiers found: '${duplicates.join(
        "', '",
      )}' - please use '?id=...' to specify unique identifiers for your cron items`,
    );
  }

  return items;
};

/**
 * Parses a list of `CronItem`s into a list of `ParsedCronItem`s, ensuring the
 * results comply with all the expectations of the `ParsedCronItem` type
 * (including those that cannot be encoded in TypeScript).
 */
export const parseCronItems = (items: CronItem[]): ParsedCronItem[] => {
  return items.map((item, idx) =>
    parseCronItem(item, `item ${idx} of parseCronItems call`),
  );
};

/**
 * Parses an individual `CronItem` into a `ParsedCronItem`, ensuring the
 * results comply with all the expectations of the `ParsedCronItem` type
 * (including those that cannot be encoded in TypeScript).
 */
export const parseCronItem = (
  cronItem: CronItem,
  source: string = "parseCronItem call",
) => {
  const {
    pattern,
    task,
    options = {} as CronItemOptions,
    payload = {},
    identifier = task,
  } = cronItem;
  const matches = CRONTAB_TIME_PARTS.exec(pattern);
  if (!matches) {
    throw new Error(`Invalid cron pattern '${pattern}' in ${source}`);
  }
  const { minutes, hours, dates, months, dows } = parseCrontabRanges(
    matches,
    source,
  );
  const item: ParsedCronItem = {
    minutes,
    hours,
    dates,
    months,
    dows,
    task,
    options,
    payload,
    identifier,
  };
  return item;
};
