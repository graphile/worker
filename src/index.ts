export * from './interfaces';

export {start, runAllJobs} from './main';
export {initWorker, Worker} from './programmatic';
