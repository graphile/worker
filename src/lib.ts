import * as assert from "assert";
import { Client, Pool } from "pg";

import { defaults } from "./config";
import { makeAddJob, makeWithPgClientFromPool } from "./helpers";
import {
  AddJobFunction,
  RunnerOptions,
  SharedOptions,
  WithPgClient,
} from "./interfaces";
import { defaultLogger, Logger, LogScope } from "./logger";
import { migrate } from "./migrate";

interface CompiledSharedOptions {
  logger: Logger;
  workerSchema: string;
  escapedWorkerSchema: string;
  maxContiguousErrors: number;
}

interface ProcessSharedOptionsSettings {
  scope?: LogScope;
}

const _sharedOptionsCache = new WeakMap<SharedOptions, CompiledSharedOptions>();
export function processSharedOptions(
  options: SharedOptions,
  { scope }: ProcessSharedOptionsSettings = {},
): CompiledSharedOptions {
  let compiled = _sharedOptionsCache.get(options);
  if (!compiled) {
    const {
      logger = defaultLogger,
      schema: workerSchema = defaults.schema,
    } = options;
    const escapedWorkerSchema = Client.prototype.escapeIdentifier(workerSchema);
    compiled = {
      logger,
      workerSchema,
      escapedWorkerSchema,
      maxContiguousErrors: defaults.maxContiguousErrors,
    };
    _sharedOptionsCache.set(options, compiled);
  }
  if (scope) {
    return {
      ...compiled,
      logger: compiled.logger.scope(scope),
    };
  } else {
    return compiled;
  }
}

export type Releasers = Array<() => void | Promise<void>>;

export async function assertPool(
  options: SharedOptions,
  releasers: Releasers,
): Promise<Pool> {
  const { logger } = processSharedOptions(options);
  assert(
    !options.pgPool || !options.connectionString,
    "Both `pgPool` and `connectionString` are set, at most one of these options should be provided",
  );
  let pgPool: Pool;
  if (options.pgPool) {
    pgPool = options.pgPool;
  } else if (options.connectionString) {
    pgPool = new Pool({
      connectionString: options.connectionString,
      max: options.maxPoolSize,
    });
    releasers.push(() => pgPool.end());
  } else if (process.env.DATABASE_URL) {
    pgPool = new Pool({
      connectionString: process.env.DATABASE_URL,
      max: options.maxPoolSize,
    });
    releasers.push(() => pgPool.end());
  } else {
    throw new Error(
      "You must either specify `pgPool` or `connectionString`, or you must make the `DATABASE_URL` environmental variable available.",
    );
  }

  pgPool.on("error", (err) => {
    /*
     * This handler is required so that client connection errors don't bring
     * the server down (via `unhandledError`).
     *
     * `pg` will automatically terminate the client and remove it from the
     * pool, so we don't actually need to take any action here, just ensure
     * that the event listener is registered.
     */
    logger.error(`PostgreSQL client generated error: ${err.message}`, {
      error: err,
    });
  });
  return pgPool;
}

export type Release = () => Promise<void>;

export async function withReleasers<T>(
  callback: (releasers: Releasers, release: Release) => Promise<T>,
): Promise<T> {
  const releasers: Releasers = [];
  const release: Release = async () => {
    await Promise.all(releasers.map((fn) => fn()));
  };
  try {
    return await callback(releasers, release);
  } catch (e) {
    try {
      await release();
    } catch (e2) {
      /* noop */
    }
    throw e;
  }
}

interface ProcessOptionsExtensions {
  pgPool: Pool;
  withPgClient: WithPgClient;
  addJob: AddJobFunction;
  release: Release;
  releasers: Releasers;
}

export interface CompiledOptions
  extends CompiledSharedOptions,
    ProcessOptionsExtensions {}

export const getUtilsAndReleasersFromOptions = async (
  options: RunnerOptions,
  settings: ProcessSharedOptionsSettings = {},
): Promise<CompiledOptions> => {
  const shared = processSharedOptions(options, settings);
  const { concurrency = defaults.concurrentJobs } = options;
  return withReleasers(
    async (releasers, release): Promise<CompiledOptions> => {
      const pgPool: Pool = await assertPool(options, releasers);
      // @ts-ignore
      const max = pgPool?.options?.max || 10;
      if (max < concurrency) {
        console.warn(
          `WARNING: having maxPoolSize (${max}) smaller than concurrency (${concurrency}) may lead to non-optimal performance.`,
        );
      }

      const withPgClient = makeWithPgClientFromPool(pgPool);

      // Migrate
      await withPgClient((client) => migrate(options, client));
      const addJob = makeAddJob(options, withPgClient);

      return {
        ...shared,
        pgPool,
        withPgClient,
        addJob,
        release,
        releasers,
      };
    },
  );
};
