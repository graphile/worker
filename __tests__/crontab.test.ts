import { CronItemOptions, ParsedCronMatch } from "../src";
import { parseCrontab } from "../src/crontab";

// 0...59
const ALL_MINUTES = Array.from(Array(60).keys());
// 0...23
const ALL_HOURS = Array.from(Array(24).keys());
// 1...31
const ALL_DATES = Array.from(Array(32).keys()).slice(1);
// 1...12
const ALL_MONTHS = Array.from(Array(13).keys()).slice(1);
// 0...6
const ALL_DOWS = [0, 1, 2, 3, 4, 5, 6];

const MINUTE = 60000;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;
const WEEK = 7 * DAY;

test("parses crontab file correctly", () => {
  const exampleCrontab = `\
# ┌───────────── UTC minute (0 - 59)
# │ ┌───────────── UTC hour (0 - 23)
# │ │ ┌───────────── UTC day of the month (1 - 31)
# │ │ │ ┌───────────── UTC month (1 - 12)
# │ │ │ │ ┌───────────── UTC day of the week (0 - 6) (Sunday to Saturday)
# │ │ │ │ │ ┌───────────── task (identifier) to schedule
# │ │ │ │ │ │    ┌────────── optional scheduling options
# │ │ │ │ │ │    │     ┌────── optional payload to merge
# │ │ │ │ │ │    │     │
# │ │ │ │ │ │    │     │
# * * * * * task ?opts {payload}

* * * * * simple
0 4 * * * every_day_at_4_am
0 4 * * 0 every_sunday_at_4_am
0 4 * * 7 every_sunday_at_4_am ?id=sunday_7




0 4 * * 2 every_tuesday_at_4_am {isTuesday: true}
*/10,7,56-59 1 1 1 1 one ?id=stuff&fill=4w3d2h1m&max=3&queue=my_queue&priority=3 {myExtraPayload:{stuff:"here with # hash char"}}
    *     *      *       *       *      lots_of_spaces     
`;
  const parsed = parseCrontab(exampleCrontab);

  expect(parsed[0].task).toEqual("simple");
  expect(parsed[0].identifier).toEqual("simple");
  const parsedCronMatch0 = (parsed[0].match as any)
    .parsedCronMatch as ParsedCronMatch;
  expect(parsedCronMatch0.minutes).toEqual(ALL_MINUTES);
  expect(parsedCronMatch0.hours).toEqual(ALL_HOURS);
  expect(parsedCronMatch0.dates).toEqual(ALL_DATES);
  expect(parsedCronMatch0.months).toEqual(ALL_MONTHS);
  expect(parsedCronMatch0.dows).toEqual(ALL_DOWS);
  expect(parsed[0].options).toEqual({ backfillPeriod: 0 });
  expect(parsed[0].payload).toEqual(null);

  expect(parsed[1].task).toEqual("every_day_at_4_am");
  expect(parsed[1].identifier).toEqual("every_day_at_4_am");
  const parsedCronMatch1 = (parsed[1].match as any)
    .parsedCronMatch as ParsedCronMatch;
  expect(parsedCronMatch1.minutes).toEqual([0]);
  expect(parsedCronMatch1.hours).toEqual([4]);
  expect(parsedCronMatch1.dates).toEqual(ALL_DATES);
  expect(parsedCronMatch1.months).toEqual(ALL_MONTHS);
  expect(parsedCronMatch1.dows).toEqual(ALL_DOWS);
  expect(parsed[1].options).toEqual({ backfillPeriod: 0 });
  expect(parsed[1].payload).toEqual(null);

  expect(parsed[2].task).toEqual("every_sunday_at_4_am");
  expect(parsed[2].identifier).toEqual("every_sunday_at_4_am");
  const parsedCronMatch2 = (parsed[2].match as any)
    .parsedCronMatch as ParsedCronMatch;
  expect(parsedCronMatch2.minutes).toEqual([0]);
  expect(parsedCronMatch2.hours).toEqual([4]);
  expect(parsedCronMatch2.dates).toEqual(ALL_DATES);
  expect(parsedCronMatch2.months).toEqual(ALL_MONTHS);
  expect(parsedCronMatch2.dows).toEqual([0]);
  expect(parsed[2].options).toEqual({ backfillPeriod: 0 });
  expect(parsed[2].payload).toEqual(null);

  expect(parsed[3].task).toEqual("every_sunday_at_4_am");
  expect(parsed[3].identifier).toEqual("sunday_7");
  const parsedCronMatch3 = (parsed[3].match as any)
    .parsedCronMatch as ParsedCronMatch;
  expect(parsedCronMatch3.minutes).toEqual([0]);
  expect(parsedCronMatch3.hours).toEqual([4]);
  expect(parsedCronMatch3.dates).toEqual(ALL_DATES);
  expect(parsedCronMatch3.months).toEqual(ALL_MONTHS);
  expect(parsedCronMatch3.dows).toEqual([0]);
  expect(parsed[3].options).toEqual({ backfillPeriod: 0 });
  expect(parsed[3].payload).toEqual(null);

  expect(parsed[4].task).toEqual("every_tuesday_at_4_am");
  expect(parsed[4].identifier).toEqual("every_tuesday_at_4_am");
  const parsedCronMatch4 = (parsed[4].match as any)
    .parsedCronMatch as ParsedCronMatch;
  expect(parsedCronMatch4.minutes).toEqual([0]);
  expect(parsedCronMatch4.hours).toEqual([4]);
  expect(parsedCronMatch4.dates).toEqual(ALL_DATES);
  expect(parsedCronMatch4.months).toEqual(ALL_MONTHS);
  expect(parsedCronMatch4.dows).toEqual([2]);
  expect(parsed[4].options).toEqual({ backfillPeriod: 0 });
  expect(parsed[4].payload).toEqual({ isTuesday: true });

  // */10,7,56-59 1 1 1 1 one ?id=stuff&fill=4w3d2h1m&max=3&queue=my_queue&priority=3 {myExtraPayload:{stuff:"here with # hash char"}}
  expect(parsed[5].task).toEqual("one");
  expect(parsed[5].identifier).toEqual("stuff");
  const parsedCronMatch5 = (parsed[5].match as any)
    .parsedCronMatch as ParsedCronMatch;
  expect(parsedCronMatch5.minutes).toEqual([
    0,
    7,
    10,
    20,
    30,
    40,
    50,
    56,
    57,
    58,
    59,
  ]);
  expect(parsedCronMatch5.hours).toEqual([1]);
  expect(parsedCronMatch5.dates).toEqual([1]);
  expect(parsedCronMatch5.months).toEqual([1]);
  expect(parsedCronMatch5.dows).toEqual([1]);
  expect(parsed[5].options).toEqual({
    backfillPeriod: 4 * WEEK + 3 * DAY + 2 * HOUR + 1 * MINUTE,
    maxAttempts: 3,
    priority: 3,
    queueName: "my_queue",
  } as CronItemOptions);
  expect(parsed[5].payload).toEqual({
    myExtraPayload: { stuff: "here with # hash char" },
  });

  expect(parsed[6].task).toEqual("lots_of_spaces");
  expect(parsed[6].identifier).toEqual("lots_of_spaces");
  const parsedCronMatch6 = (parsed[6].match as any)
    .parsedCronMatch as ParsedCronMatch;
  expect(parsedCronMatch6.minutes).toEqual(ALL_MINUTES);
  expect(parsedCronMatch6.hours).toEqual(ALL_HOURS);
  expect(parsedCronMatch6.dates).toEqual(ALL_DATES);
  expect(parsedCronMatch6.months).toEqual(ALL_MONTHS);
  expect(parsedCronMatch6.dows).toEqual(ALL_DOWS);
  expect(parsed[6].options).toEqual({ backfillPeriod: 0 });
  expect(parsed[6].payload).toEqual(null);

  expect(parsed).toMatchSnapshot();
});

describe("gives error on syntax error", () => {
  test("too few parameters", () => {
    expect(() =>
      parseCrontab(`\
* * * * too_few_parameters
`),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Could not process line '1' of crontab: '* * * * too_few_parameters'"`,
    );
  });

  test("invalid command (two parts)", () => {
    expect(() =>
      parseCrontab(`\
* * * * * two tasks
`),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid command specification in line 1 of crontab."`,
    );
  });

  test("range exceeded", () => {
    expect(() =>
      parseCrontab(`\
1,60 * * * * out_of_range
`),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Too large value '60' in minutes range in line 1 of crontab: expected values in the range 0-59."`,
    );
  });

  test("invalid wildcard divisor", () => {
    expect(() =>
      parseCrontab(`\
*/0 * * * * division_by_zero
`),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Invalid wildcard expression '*/0' in minutes range in line 1 of crontab: divisor '0' expected to be greater than zero"`,
    );
  });

  test("unknown option", () => {
    expect(() =>
      parseCrontab(`\
* * * * * invalid_options ?unknown=3
`),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Options on line 1 of crontab contains unsupported key 'unknown'; supported keys are: 'id', 'fill', 'max', 'queue', 'priority'."`,
    );
  });

  test("invalid JSON5 syntax", () => {
    expect(() =>
      parseCrontab(`\
* * * * * json_syntax_error {invalidJson=true}
`),
    ).toThrowErrorMatchingInlineSnapshot(
      `"Failed to parse JSON5 payload on line 1 of crontab: JSON5: invalid character '=' at 1:13"`,
    );
  });
});
