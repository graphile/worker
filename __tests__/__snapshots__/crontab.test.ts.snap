// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses crontab file correctly 1`] = `
Array [
  Object {
    "dates": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
      26,
      27,
      28,
      29,
      30,
      31,
    ],
    "dows": Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
    ],
    "hours": Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
    ],
    "identifier": "simple",
    "minutes": Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
      26,
      27,
      28,
      29,
      30,
      31,
      32,
      33,
      34,
      35,
      36,
      37,
      38,
      39,
      40,
      41,
      42,
      43,
      44,
      45,
      46,
      47,
      48,
      49,
      50,
      51,
      52,
      53,
      54,
      55,
      56,
      57,
      58,
      59,
    ],
    "months": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
    ],
    "options": Object {
      "backfillPeriod": 0,
      "maxAttempts": undefined,
      "priority": undefined,
      "queueName": undefined,
    },
    "payload": null,
    "task": "simple",
  },
  Object {
    "dates": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
      26,
      27,
      28,
      29,
      30,
      31,
    ],
    "dows": Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
    ],
    "hours": Array [
      4,
    ],
    "identifier": "every_day_at_4_am",
    "minutes": Array [
      0,
    ],
    "months": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
    ],
    "options": Object {
      "backfillPeriod": 0,
      "maxAttempts": undefined,
      "priority": undefined,
      "queueName": undefined,
    },
    "payload": null,
    "task": "every_day_at_4_am",
  },
  Object {
    "dates": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
      26,
      27,
      28,
      29,
      30,
      31,
    ],
    "dows": Array [
      0,
    ],
    "hours": Array [
      4,
    ],
    "identifier": "every_sunday_at_4_am",
    "minutes": Array [
      0,
    ],
    "months": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
    ],
    "options": Object {
      "backfillPeriod": 0,
      "maxAttempts": undefined,
      "priority": undefined,
      "queueName": undefined,
    },
    "payload": null,
    "task": "every_sunday_at_4_am",
  },
  Object {
    "dates": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
      26,
      27,
      28,
      29,
      30,
      31,
    ],
    "dows": Array [
      0,
    ],
    "hours": Array [
      4,
    ],
    "identifier": "sunday_7",
    "minutes": Array [
      0,
    ],
    "months": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
    ],
    "options": Object {
      "backfillPeriod": 0,
      "maxAttempts": undefined,
      "priority": undefined,
      "queueName": undefined,
    },
    "payload": null,
    "task": "every_sunday_at_4_am",
  },
  Object {
    "dates": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
      26,
      27,
      28,
      29,
      30,
      31,
    ],
    "dows": Array [
      2,
    ],
    "hours": Array [
      4,
    ],
    "identifier": "every_tuesday_at_4_am",
    "minutes": Array [
      0,
    ],
    "months": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
    ],
    "options": Object {
      "backfillPeriod": 0,
      "maxAttempts": undefined,
      "priority": undefined,
      "queueName": undefined,
    },
    "payload": Object {
      "isTuesday": true,
    },
    "task": "every_tuesday_at_4_am",
  },
  Object {
    "dates": Array [
      1,
    ],
    "dows": Array [
      1,
    ],
    "hours": Array [
      1,
    ],
    "identifier": "stuff",
    "minutes": Array [
      0,
      7,
      10,
      20,
      30,
      40,
      50,
      56,
      57,
      58,
      59,
    ],
    "months": Array [
      1,
    ],
    "options": Object {
      "backfillPeriod": 2685660000,
      "maxAttempts": 3,
      "priority": 3,
      "queueName": "my_queue",
    },
    "payload": Object {
      "myExtraPayload": Object {
        "stuff": "here with # hash char",
      },
    },
    "task": "one",
  },
  Object {
    "dates": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
      26,
      27,
      28,
      29,
      30,
      31,
    ],
    "dows": Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
    ],
    "hours": Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
    ],
    "identifier": "lots_of_spaces",
    "minutes": Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
      26,
      27,
      28,
      29,
      30,
      31,
      32,
      33,
      34,
      35,
      36,
      37,
      38,
      39,
      40,
      41,
      42,
      43,
      44,
      45,
      46,
      47,
      48,
      49,
      50,
      51,
      52,
      53,
      54,
      55,
      56,
      57,
      58,
      59,
    ],
    "months": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
    ],
    "options": Object {
      "backfillPeriod": 0,
      "maxAttempts": undefined,
      "priority": undefined,
      "queueName": undefined,
    },
    "payload": null,
    "task": "lots_of_spaces",
  },
]
`;
